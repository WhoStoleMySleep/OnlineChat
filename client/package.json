{
  "name": "client",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@apollo/client": "^3.6.9",
    "@apollo/react-hoc": "^4.0.0",
    "@faker-js/faker": "^7.3.0",
    "@reduxjs/toolkit": "^1.8.3",
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^13.3.0",
    "@testing-library/user-event": "^14.2.1",
    "@types/jest": "^27.5.2",
    "@types/node": "^17.0.45",
    "@types/react": "^18.0.15",
    "@types/react-dom": "^18.0.6",
    "classnames": "^2.3.1",
    "date-fns": "^2.29.1",
    "mathjs": "^10.6.4",
    "push.js": "^1.0.12",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-redux": "^8.0.2",
    "react-scripts": "5.0.1",
    "sass": "^1.53.0",
    "typescript": "^4.7.4",
    "use-sound": "^4.0.1",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test --coverage",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^5.30.6",
    "@typescript-eslint/parser": "^5.30.6",
    "babel-plugin-istanbul": "^6.1.1",
    "eslint": "^8.20.0",
    "eslint-config-airbnb": "^19.0.4",
    "eslint-plugin-import": "^2.26.0",
    "eslint-plugin-jest": "^26.6.0",
    "eslint-plugin-jsx-a11y": "^6.6.0",
    "eslint-plugin-react": "^7.30.1",
    "eslint-plugin-react-hooks": "^4.6.0",
    "react-test-renderer": "^18.2.0"
  }
}
