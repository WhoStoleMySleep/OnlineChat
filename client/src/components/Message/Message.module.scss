.message {
  $self: &;
  
  margin: 20px 40px 15px 20px;
  width: fit-content;

  &.me-author {
    margin: 0 20px 15px auto;
    margin-left: auto;
    text-align: right;

    #{$self}__content {
      border-radius: 10px 10px 0 10px;
      margin-left: auto;
      margin-right: 0;

      &::before {
        border-radius: 0 100% 30% 0;
        left: auto;
        right: -10px;
      }
    }

    #{$self}__author {
      pointer-events: none;
    }

    #{$self}__date {
      text-align: right;
    }
  }

  &__content {
    background: rgba(255, 255, 255, 0.2);
    border-radius: 10px 10px 10px 0;
    margin-right: auto;
    max-width: 400px;
    padding: 5px 20px;
    position: relative;
    text-align: left;
    width: fit-content;
    box-sizing: border-box;
  }
  
  &__author,
  &__text,
  &__date {
    word-wrap: break-word;
    color: #fff;
    pointer-events: none;
  }
  
  &__author {
    font-size: 23px;
    font-weight: 600;
    margin-bottom: 10px;
  }
  
  &__text {
    background: none;
    border: none;
    font-size: 20px;
    max-width: 100%;
    outline: none;
    overflow: hidden;
    resize: none;
  }
  
  textarea {
    pointer-events: all;
  }
  
  &__date {
    font-size: 12px;
    margin-top: 8px;
    opacity: 0.8;
  }
}

@media all and (max-width: 900px) {
  .message {
    max-width: 40%;

    &__content {
      max-width: 100%;
    }
  }
}

@media all and (max-width: 620px) {
  .message {
    margin: 20px 40px 0 5px;
    max-width: 60%;

    &::before {
      height: 4px;
      left: -5px;
      width: 5px;
    }

    &.me-author {
      margin: 20px 5px 40px auto;

      &::before {
        right: -5px;
      }
    }
  }
}