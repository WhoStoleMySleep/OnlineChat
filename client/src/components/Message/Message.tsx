import React, { useState } from 'react';
import formatDistanceToNow from 'date-fns/formatDistanceToNow';
import classNames from 'classnames';
import useContextMenu from '../../hooks/useContextMenu/useContextMenu';
import useRetainUpdatedValue from '../../hooks/useRetainUpdatedValue/useRetainUpdatedValue';
import contextMenuStyles from '../ContextMenu/ContextMenu.module.scss';
import styles from './Message.module.scss';

function Message(
  props: {
    id: string,
    MeAuthor: string,
    receivedAuthor: string,
    text: string,
    date: string
  }
) {
  const {
    id, MeAuthor, receivedAuthor, text, date
  } = props;

  const [editId, setEditId] = useState('');

  const { onBlur } = useRetainUpdatedValue('onBlur');
  const { onContextMenu } = useContextMenu(
    contextMenuStyles['context-menu'],
    contextMenuStyles.active,
    contextMenuStyles.reversed,
  );

  function autoSize(event: { target: HTMLElement; }) {
    const element = event.target;

    element.style.height = 'auto';
    element.style.height = `${element.scrollHeight - 4}px`;
  }

  const confirmWithEnter = (event: { key: string; target: { blur: () => void; }; }) => {
    if (event.key === 'Enter') {
      event.target.blur();
    }
  };

  return (
    <li
      data-testid="Message"
      className={classNames(styles.message, {
        [styles['me-author']]: MeAuthor === receivedAuthor,
        [styles.edit]: editId === id
      })}
      onContextMenu={
        (event) => (
          MeAuthor === receivedAuthor
            ? onContextMenu(event, setEditId, id)
            : null
        )
      }
      onTouchStart={
        (event) => (
          MeAuthor === receivedAuthor
            ? onContextMenu(event, setEditId, id)
            : null
        )
      }
    >
      <div className={styles.message__content}>
        <p className={styles.message__author}>{receivedAuthor}</p>
        {editId !== id
          ? <p className={styles.message__text}>{text}</p>
          : (
            <textarea
              defaultValue={text}
              className={styles.message__text}
              onChange={autoSize}
              onKeyDown={(event) => confirmWithEnter(event)}
              onBlur={(event) => onBlur(event, id, setEditId)}
            />
          )}
      </div>
      {date
        && (
        <p className={styles.message__date}>
          {formatDistanceToNow(new Date(date), { addSuffix: true })}
        </p>
        )}
    </li>
  );
}

export default Message;
