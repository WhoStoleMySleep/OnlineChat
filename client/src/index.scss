body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  background: linear-gradient(
    120deg,
    rgba(23, 190, 187, 1),
    rgba(240, 166, 202, 1)
  );
  background-repeat: no-repeat;
  background-size: 100% 100%;
  height: 100vh;
  overflow-x: hidden;
}
