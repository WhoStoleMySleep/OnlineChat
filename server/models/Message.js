const { model, Schema } = require('mongoose');

const messageSchema = new Schema({
  text: String,
  author: String,
});

module.exports = model('Message', messageSchema);
