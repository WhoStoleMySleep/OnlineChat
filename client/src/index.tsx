import {
  ApolloClient,
  ApolloProvider,
  HttpLink,
  InMemoryCache,
  split,
} from '@apollo/client';
import React from 'react';
import { createRoot } from 'react-dom/client';
import { Provider } from 'react-redux';
import App from './App';
import './index.scss';
import store from './redux/store';
import { WebSocketLink } from '@apollo/client/link/ws';
import { getMainDefinition } from '@apollo/client/utilities';

const httpLink = new HttpLink({
  uri: 'https://adasdasdasdasdasdsdfsdfsdfsdfs.herokuapp.com/graphql',
});

const wsLink = new WebSocketLink({
  uri: 'wss://adasdasdasdasdasdsdfsdfsdfsdfs.herokuapp.com/graphql',
  options: {
    reconnect: true,
  },
});

const splitLink = split(
  ({ query }) => {
    const definition = getMainDefinition(query);
    return (
      definition.kind === 'OperationDefinition' &&
      definition.operation === 'subscription'
    );
  },
  wsLink,
  httpLink
);

const client = new ApolloClient({
  link: splitLink,
  cache: new InMemoryCache(),
});

const container = document.getElementById('root')!;
const root = createRoot(container);

root.render(
  <React.StrictMode>
    <ApolloProvider client={client}>
      <Provider store={store}>
        <App />
      </Provider>
    </ApolloProvider>
  </React.StrictMode>
);
