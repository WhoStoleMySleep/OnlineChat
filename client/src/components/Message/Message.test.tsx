import '@testing-library/jest-dom';
import {
  render, screen
} from '@testing-library/react';
import React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';
import Message from './Message';
import { MESSAGES_QUERY } from '../../GraphQl.queries';

const date = new Date().toString();

const mocks = [{
  request: {
    query: MESSAGES_QUERY,
  },
  result: {
    data: {
      id: '123', text: '123', author: '321', date
    }
  }
}];

describe('Message', () => {
  it('Message snapshot', async () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Message
          id="123"
          MeAuthor="123"
          receivedAuthor="321"
          text="123"
          date={date}
        />
      </MockedProvider>
    );

    const message = screen.getByTestId('Message');

    expect(message).toMatchSnapshot();
  });

  it('asd', () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Message
          id="123"
          MeAuthor="123"
          receivedAuthor="321"
          text="123"
          date=""
        />
      </MockedProvider>
    );

    const message = screen.getByTestId('Message');
    const messageDate = message.querySelector('.message__date');

    expect(messageDate).toBeNull();
  });
});
