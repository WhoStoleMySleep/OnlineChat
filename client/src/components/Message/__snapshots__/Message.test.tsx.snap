// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Message snapshot 1`] = `
<li
  class="message"
  data-testid="Message"
>
  <div
    class="message__content"
  >
    <p
      class="message__author"
    >
      321
    </p>
    <p
      class="message__text"
    >
      123
    </p>
  </div>
  <p
    class="message__date"
  >
    less than a minute ago
  </p>
</li>
`;
